src/Compilers/VisualBasic/Portable/Symbols/ObsoleteAttributeHelpers.vb

```diff
@@ -72,11 +72,42 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Debug.Assert(context IsNot Nothing)
             Debug.Assert(symbol IsNot Nothing)
 
+            Dim namedType = TryCast(symbol, NamedTypeSymbol)
+            If namedType IsNot Nothing AndAlso
+                IsExperimentalSymbol(namedType) Then
+
+                ' Skip for System.Diagnostics.CodeAnalysis.ExperimentalAttribute to mitigate cycles
+                Return ObsoleteDiagnosticKind.NotObsolete
+            End If
+
+            Dim method = TryCast(symbol, MethodSymbol)
+            If method IsNot Nothing AndAlso
+                    method.MethodKind = MethodKind.Constructor AndAlso
+                    IsExperimentalSymbol(method.ContainingType) Then
+
+                ' Skip for constructors of System.Diagnostics.CodeAnalysis.ExperimentalAttribute to mitigate cycles
+                Return ObsoleteDiagnosticKind.NotObsolete
+            End If
+
+            If symbol.ContainingModule?.ObsoleteKind = ObsoleteAttributeKind.Experimental OrElse
+                symbol.ContainingAssembly?.ObsoleteKind = ObsoleteAttributeKind.Experimental Then
+
+                Return GetDiagnosticKind(context, forceComplete, getStateFromSymbol:=Function(s) s.ExperimentalState)
+            End If
+
             Select Case symbol.ObsoleteKind
                 Case ObsoleteAttributeKind.None
+                    If symbol.ContainingModule?.ObsoleteKind = ObsoleteAttributeKind.Uninitialized OrElse
```

Comment:
>symbol.ContainingModule? [](http://example.com/codeflow?start=23&length=24)

We're allowing for `ContainingModule` and `ContainingAssembly` to be `null` here but not in C#. Is there a reason for the difference?

Reply:
That's an existing difference. Here's the relevant comment on the `ContainingModule` and `ContainingAssembly` APIs (although they weren't super helpful for me to understand...):
```vb
        ''' Returns the module [assembly] containing this symbol. If this symbol is shared
        ''' across multiple modules [assemblies], or doesn't belong to a module [assembly], returns Nothing.
``` 

I don't know about the sharing scenario. Maybe that's a problem for [Experimental] (seems like we'd still want to enforce). I'll dig a bit more.